app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/PlayBillingRepository.kt

```diff
@@ -0,0 +1,164 @@
+package pm.gnosis.heimdall.data.repositories.impls
+
+import android.app.Activity
+import android.content.Context
+import com.android.billingclient.api.*
+import com.gojuno.koptional.Optional
+import com.gojuno.koptional.toOptional
+import io.reactivex.Single
+import io.reactivex.SingleEmitter
+import io.reactivex.SingleOnSubscribe
+import io.reactivex.rxkotlin.subscribeBy
+import pm.gnosis.heimdall.data.repositories.BillingRepository
+import pm.gnosis.heimdall.data.repositories.models.UserPurchase
+import pm.gnosis.heimdall.helpers.SetStore
+import pm.gnosis.svalinn.common.di.ApplicationContext
+import timber.log.Timber
+import java.lang.ref.WeakReference
+import java.util.*
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class PlayBillingRepository @Inject constructor(
+    @ApplicationContext private val context: Context
+) : BillingRepository, PurchasesUpdatedListener {
+
+    private val purchaseStore = SetStore<String>()
+
+    private val billingClient = BillingClient.newBuilder(context).setListener(this).build()
+
+    private val pendingActionsLock = Any()
+    private val pendingActions = ArrayList<BillingAction>()
+
+    private var isConnected = false
+
+    override fun init() {
+        updatePurchases()
+    }
+
+    private fun updatePurchases() {
+        billingAction {
+            purchaseStore.replaceAll(
+                billingClient.queryPurchases(BillingClient.SkuType.SUBS).purchasesList
+                    .map { it.sku }
+            )
+        }.subscribeBy(onError = Timber::e)
+    }
+
+    override fun checkSubscription(id: String): Single<Optional<UserPurchase>> =
+        billingAction {
+            billingClient.queryPurchases(BillingClient.SkuType.SUBS)
+                .purchasesList
+                .find { it.sku == id }
+                ?.let { UserPurchase(it.purchaseToken, it.orderId) }
+                .toOptional()
+        }
+
+    override fun launchSubscribeFlow(activity: Activity, productId: String): Single<Boolean> {
+        val weakActivity = WeakReference(activity)
+        return billingAction {
+            weakActivity.get()?.let {
+                billingClient.launchBillingFlow(
+                    it, BillingFlowParams.newBuilder()
+                        .setType(BillingClient.SkuType.SUBS)
+                        .setSku(productId)
+                        .build()
+                )
+                true
+            } ?: run {
+                // Default to error if no activity found
+                false
+            }
+        }
+    }
+
+    override fun onPurchasesUpdated(billingResponse: Int, purchases: MutableList<Purchase>?) {
+        updatePurchases()
```

Comment:
Maybe emitting an event to the stream might be better than creating a new disposable.